lib/recursive-methods.rb

```diff
@@ -1,49 +1,72 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 0
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1 
+  rev_s = reverse(s[1..-1])
+  rev_s += s[0]
+  return rev_s
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def reverse_inplace(s, first = 0, last = s.length - 1)
+  return s if first >= last
+  temp = s[first]
+  s[first] = s[last]
+  s[last] = temp
+  
+  return reverse_inplace(s, first + 1, last - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n <= 0
+  return 2 if n == 1
+  return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def nested(s, i = 0)
+  string = s.split("()")
+  return false if s.length.odd?
+  return true if s.empty? 
+  
+  return true if s.length.even? && string[i].length == string[i+1].length
+  return false if s.length.even? && string[i].length != string[i+1].length
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
-end
+# Time complexity: O(n)
+# Space complexity: O(n)
+def search(array, value, index = 0)
+  return false if index >= array.length
+  return true if array[index] == value
+  return search(array, value, index + 1)
+end 
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length <= 1
+  return false if s[0] != s[-1]
+  return is_palindrome(s[1...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def digit_match(n, m)
-    raise NotImplementedError, "Method not implemented"
-end
\ No newline at end of file
+  return 1 if n == 0 && m == 0
+  return 0 if n == 0 || m == 0
+  return 1 if n.to_s[-1] == m.to_s[-1]
+end
```

Comment:
You're only checking the last digit here.

Maybe use `if n % 10 == m % 10` to compare the last digit and then divide by 10 to repeat with the remaining digits.